File: src/main/java/Parser.java
import java.io.IOException;

public class Parser {
   private Lexer lex;
   private Token tok;

   public Parser(Lexer lex) throws IOException {
      this.lex = lex;
      this.tok = lex.getToken();
   }

   private void error(Token.T[] expected) {
      StringBuilder b = new StringBuilder();
      if (expected.length > 0) {
         b.append(expected[0]);
         for (int i = 1; i < expected.length; i++)
            b.append(", ").append(expected[i]);
      }
      ErrorMsg.error(tok.line, tok.col, "syntax error", "expecting " + b.toString() + ", found " + tok.type);
      System.exit(3);
   }

   private void advance() throws IOException {
      tok = lex.getToken();
   }

   private void eat(Token.T t) throws IOException {
      if (tok.type == t)
         advance();
      else
         error(new Token.T[]{t});
   }

   private void S() throws IOException {
      switch (tok.type) {
         case IF:
            eat(Token.T.IF);
            E();
            eat(Token.T.THEN);
            S();
            eat(Token.T.ELSE);
            S();
            break;
         case BEGIN:
            eat(Token.T.BEGIN);
            S();
            L();
            break;
         case PRINT:
            eat(Token.T.PRINT);
            E();
            break;
         default:
            error(new Token.T[]{Token.T.IF, Token.T.BEGIN, Token.T.PRINT});
      }
   }

   private void L() throws IOException {
      switch (tok.type) {
         case END:
            eat(Token.T.END);
            break;
         case SEMI:
            eat(Token.T.SEMI);
            S();
            L();
            break;
         default:
            error(new Token.T[]{Token.T.END, Token.T.SEMI});
      }
   }

   private void E() throws IOException {
      eat(Token.T.NUM);
      eat(Token.T.EQ);
      eat(Token.T.NUM);
   }

   public void parse() throws IOException {
      S();
      eat(Token.T.EOF);
   }
}

File: ocaml/src/parser.ml
(* Analisador sintático para a linguagem da gramática 3.11 do livro do
Appel *)

open Token

type parsebuf = { lexbuf : Lexing.lexbuf;
                  next_token : Lexing.lexbuf -> token;
                  mutable tok : token
                }

let parsebuf_of_lexbuf lb token =
  { lexbuf = lb;
    next_token = token;
    tok = token lb
  }

let rec intersperse sep = function
  | [] -> ""
  | [s] -> s
  | s::others -> s ^ sep ^ intersperse sep others

exception Syntax of string

let error tokens_expected token_found =
  raise
    (Syntax
       (Printf.sprintf
          "expected %s, but found %s"
          (intersperse ", " (List.map show_token tokens_expected))
          (show_token token_found)))

let advance pb =
  pb.tok <- pb.next_token pb.lexbuf

let eat pb tok =
  if same_token_type pb.tok tok then
    advance pb
  else
    error [tok] pb.tok

let rec s pb =
  match pb.tok with
  | TIf -> advance pb; e pb; eat pb TThen; s pb; eat pb TElse; s pb
  | TBegin -> advance pb; s pb; l pb
  | TPrint -> advance pb; e pb
  | _ -> error [TIf; TBegin; TPrint] pb.tok

and l pb =
  match pb.tok with
  | TEnd -> advance pb
  | TSemi -> advance pb; s pb; l pb
  | _ -> error [TEnd; TSemi] pb.tok

and e pb =
  eat pb (TNum 0); eat pb TEq; eat pb (TNum 0)

and prog pb =
  s pb; eat pb TEOF
